zookeeper-server/src/main/java/org/apache/zookeeper/server/ResponseCache.java

```diff
@@ -0,0 +1,84 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.zookeeper.server;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import org.apache.jute.Record;
+import org.apache.zookeeper.data.Stat;
+
+@SuppressWarnings("serial")
+public class ResponseCache {
+    private static final int DEFAULT_RESPONSE_CACHE_SIZE = 400;
+
+    private static class Entry {
+        public Stat stat;
+        public byte[] data;
+    }
+
+    private Map<String, Entry> cache = Collections.synchronizedMap(
```

Comment:
Can LRUcache compose the LinkedHashMap rather extending? Can we use read-write locks for better synchronization? The read lock on reads will give better results rather locking on the map.

Reply:
Would be good if DEFAULT_INITIAL_CAPACITY of cache also defined

Reply:
The advice to use Collections::synchronizedMap is from the java docs for LinkedHashMap (https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html). I'm happier using this pattern for now than changing it up as we should not be building our own caching logic. 

One alternative would be to use a guava cache.